### DISCUSSION TITLE
gTTs Attribute error

### DISCUSSION THREAD

**Discussion by pookieebear** [ultralytics/ultralytics#10286] [General]
Hi. I am trying to intergrate gTTs into yolo and I'm getting an Attribute error 'list' object has no attribute 'predict'

Here's my code `import cv2
from gtts import gTTS
import os
import pygame
from ultralytics import YOLO

def convert_labels_to_text(labels):
    if labels:
        return ", ".join(labels)
    else:
        return "No objects detected"

class YOLOWithLabels(YOLO):
    def __call__(self, frame):
        results = super().__call__(frame)
        labels = results.predict[0].get_field("labels").tolist()
        annotated_frame = results.render()
        return annotated_frame, labels

pygame.mixer.init()

cap = cv2.VideoCapture(0)
model = YOLOWithLabels('yolov8n.pt')

while cap.isOpened():
    success, frame = cap.read()

    if success:
        annotated_frame, labels = model(frame)

        message = convert_labels_to_text(labels)

        tts_engine = gTTS(text=message, lang='en', slow=False)  # Initialize gTTS with the message

        # Save the speech output to a file
        tts_engine.save("output.mp3")

        # Play the speech output
        pygame.mixer.music.load("output.mp3")
        pygame.mixer.music.play()

        cv2.putText(annotated_frame, message, (10, 30), cv2.FONT_HERSHEY_SIMPLEX, 1, (0, 255, 0), 2)
        cv2.imshow("YOLOv8 Inference", annotated_frame)

        if cv2.waitKey(1) & 0xFF == ord("q"):
            break

    else:
        break

cap.release()
cv2.destroyAllWindows()
pygame.mixer.quit()
`

**Comment by glenn-jocher** [comment]
Hello! It looks like you’re encountering an error because the `results` object returned by a YOLO model call doesn't directly have a `predict` attribute. You might be looking for the `results.xyxy[0]` to access detections or `results.pandas().xyxy[0]` to access them in a pandas DataFrame which includes labels.

Here's a bit of revised code for the part where you're attempting to get the labels:

```python
def __call__(self, frame):
    results = super().__call__(frame)  # Call the base YOLO model prediction
    labels = results.names  # Get class names
    detected_classes = results.pandas().xyxy[0]['name'].tolist()  # Get detected classes' names
    annotated_frame = results.render()[0]  # Get annotated frame
    return annotated_frame, detected_classes
```
In this snippet, `detected_classes` will give you a list of labels (class names) detected in the frame, which seems to be what you're after for text-to-speech conversion. 

Also, ensure you're managing resources well. For example, continually saving and playing `output.mp3` in a loop might cause delays or unexpected behavior. Consider using a more direct way to output speech if real-time performance is critical.

Hope this helps you move forward with integrating gTTs into your YOLO project! 🚀

**Reply by pookieebear** [reply]
I have adjusted the code accordingly and now getting new error 
![image](https://github.com/ultralytics/ultralytics/assets/141935848/a1ea3618-f367-42a7-9f24-a67a9664e140)

**Reply by pookieebear** [reply]
![Uploading image.png…]()

**Reply by glenn-jocher** [reply]
It seems like your image didn't upload correctly. Could you please try uploading it again? That way, I can better understand the issue and assist you more effectively! 😊
